bin/wasm-node/javascript/src/index-browser.ts

```diff
@@ -121,23 +113,294 @@ function trustedBase64Decode(base64: string): Uint8Array {
+          config.onStreamClose(dataChannelId);
+      };
+
+      dataChannel.onmessage = (m) => {
+          // The `data` field is an `ArrayBuffer`.
+          config.onMessage(new Uint8Array(m.data), dataChannelId);
+      }
+
+      dataChannels.set(dataChannelId, dataChannel);
+    }
+
+    // It is possible for the browser to use multiple different certificates.
+    // In order for our local certificate to be deterministic, we need to generate it manually and
+    // set it explicitly as part of the configuration.
+    // According to <https://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-generatecertificate>,
+    // browsers are guaranteed to support `{ name: "ECDSA", namedCurve: "P-256" }`.
+    RTCPeerConnection.generateCertificate({ name: "ECDSA", namedCurve: "P-256" } as EcKeyGenParams).then((localCertificate) => {
+      if (cancelOpening)
+        return;
+
+      // We need to build the multihash corresponding to the local certificate.
+      let localTlsCertificateMultihash: Uint8Array | null = null;
+      for (const { algorithm, value } of localCertificate.getFingerprints()) {
+        if (algorithm === 'sha-256') {
+          localTlsCertificateMultihash = new Uint8Array(34);
+          localTlsCertificateMultihash.set([0x12, 32], 0);
+          localTlsCertificateMultihash.set(value!.split(':').map((s) => parseInt(s, 16)), 2);
+          break;
+        }
+      }
+      if (localTlsCertificateMultihash === null) {
+        // Because we've already returned from the `connect` function at this point, we pretend
+        // that the connection has failed to open.
+        config.onConnectionClose('Failed to obtain the browser certificate fingerprint');
+        return;
+      }
+
+      // Create a new WebRTC connection.
+      pc = new RTCPeerConnection({ certificates: [localCertificate] });
+
+      // `onconnectionstatechange` is used to detect when the connection has closed or has failed
+      // to open.
+      // Note that smoldot will think that the connection is open even when it is still opening.
+      // Therefore we don't care about events concerning the fact that the connection is now fully
+      // open.
+      pc.onconnectionstatechange = (_event) => {
+        if (pc!.connectionState == "closed" || pc!.connectionState == "disconnected" || pc!.connectionState == "failed") {
+          config.onConnectionClose("WebRTC state transitioned to " + pc!.connectionState);
+
+          pc!.onconnectionstatechange = null;
+          pc!.onnegotiationneeded = null;
+          pc!.ondatachannel = null;
+
+          for (const channel of Array.from(dataChannels.values())) {
+            channel.onopen = null;
+            channel.onerror = null;
+            channel.onclose = null;
+            channel.onmessage = null;
+          }
+
+          pc!.close();  // Unclear whether this is necessary, but it doesn't hurt to do so.
+
+          for (const channel of Array.from(dataChannels.values())) {
```

Comment:
this maybe redundant. https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/close

"Calling this method terminates the RTCPeerConnection's ICE agent, ending any ongoing ICE processing and **any active streams.**"